src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,355 @@
+pub struct RefreshTokenRequest {
+    url: String,
+    authentication: ClientAuthentication,
+    headers: http::HeaderMap,
+    refresh_token: String,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use httptest::{Expectation, Server, matchers::*, responders::*};
+    type TestResult = std::result::Result<(), Box<dyn std::error::Error>>;
+
+    #[tokio::test]
+    async fn exchange_token() -> TestResult {
+        let client_auth = ClientAuthentication {
+            auth_style: ClientAuthStyle::InHeader,
+            client_id: Some("client_id".to_string()),
+            client_secret: Some("supersecret".to_string()),
+        };
+        let response_body = r#"{"access_token":"an_example_token","issued_token_type":"urn:ietf:params:oauth:token-type:access_token","token_type":"Bearer","expires_in":3600,"scope":"https://www.googleapis.com/auth/cloud-platform"}"#;
+
+        let server = Server::run();
+        server.expect(
+            Expectation::matching(all_of![
+                request::method_path("POST", "/token"),
+                request::body(url_decoded(contains((
+                    "grant_type",
+                    TOKEN_EXCHANGE_GRANT_TYPE
+                )))),
+                request::body(url_decoded(contains(("subject_token", "an_example_token")))),
+                request::body(url_decoded(contains((
+                    "requested_token_type",
+                    ACCESS_TOKEN_TYPE
+                )))),
+                request::body(url_decoded(contains((
+                    "subject_token_type",
+                    JWT_TOKEN_TYPE
+                )))),
+                request::body(url_decoded(contains((
+                    "audience",
+                    "32555940559.apps.googleusercontent.com"
+                )))),
+                request::body(url_decoded(contains((
+                    "scope",
+                    "https://www.googleapis.com/auth/devstorage.full_control"
+                )))),
+                request::headers(contains((
+                    "authorization",
+                    "Basic Y2xpZW50X2lkOnN1cGVyc2VjcmV0"
+                ))),
+                request::headers(contains((
+                    "content-type",
+                    "application/x-www-form-urlencoded"
+                ))),
+            ])
+            .respond_with(status_code(200).body(response_body)),
+        );
+
+        let url = server.url("/token");
+        let mut headers = http::HeaderMap::new();
+        headers.insert(
+            http::header::CONTENT_TYPE,
+            http::HeaderValue::from_static("application/x-www-form-urlencoded"),
+        );
+        let token_req = ExchangeTokenRequest {
+            url: url.to_string(),
+            headers: headers,
+            authentication: client_auth,
+            audience: Some("32555940559.apps.googleusercontent.com".to_string()),
+            scope: ["https://www.googleapis.com/auth/devstorage.full_control".to_string()].to_vec(),
```

Comment:
Its weird to see the requested scope and granted scope being different, especially the granted scope is much more powerful than the requested one.